glean-core/src/upload/mod.rs

```diff
@@ -440,11 +529,50 @@ mod test {
         assert_eq!(upload_manager.get_upload_task(false), PingUploadTask::Done);
     }
 
+    #[test]
+    fn test_limits_the_number_of_pings_per_minute() {
+        // Create a new upload_manager
+        let dir = tempfile::tempdir().unwrap();
+        let mut upload_manager = PingUploadManager::new(dir.path(), false);
+
+        // Wait for processing of pending pings directory to finish.
+        while upload_manager.get_upload_task(false) == PingUploadTask::Wait {
+            thread::sleep(Duration::from_millis(10));
+        }
+
+        // Enqueue a ping multiple times
+        let n = MAX_PING_SENDS_PER_MINUTE;
+        for _ in 0..n {
+            upload_manager.enqueue_ping(DOCUMENT_ID, PATH, json!({}));
+        }
+
+        // Verify a request is returned for each submitted ping
+        for _ in 0..n {
+            match upload_manager.get_upload_task(false) {
+                PingUploadTask::Upload(_) => {}
+                _ => panic!("Expected upload manager to return the next request!"),
+            }
+        }
+
+        // Enqueue just one more ping...
+        upload_manager.enqueue_ping(DOCUMENT_ID, PATH, json!({}));
+
+        // Verify that we are indeed told to wait because we are at capacity
+        assert_eq!(PingUploadTask::Wait, upload_manager.get_upload_task(false));
+
+        thread::sleep(Duration::from_millis((1000 * SECONDS_IN_A_MINUTE) as u64));
```

Comment:
I know this is way too much time for a test to wait. I'll think of another way to test it before finishing this.

Reply:
Maybe you can make the wait time customizable? We'll need that (in the future) so that products could customize it.